MCO-88: Add runbook for MCCDrainError alert (#214)
openshift/runbooks: alerts/machine-config-operator/MachineConfigControllerDrainError.md

The `machine-config-controller` pod is the central point of management for incoming updates to 
machines.

For the following command, replace the $CONTROLLERPOD variable with the name of your own `machine-config-controller` pod name. 

```console
oc -n openshift-machine-config-operator logs $CONTROLLERPOD -c machine-config-controller
```
When the MCO starts draining a node, the Machine Config Controller (MCC) records the following log entry:

```console
1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: initiating drain
```
The MCC then logs the name of each pod that is drained from the node.

```console
  1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Evicted pod <namespace>/<pod-name>
```

If the MCO/MCC is unable to drain a pod after 1m30s, the MCC logs the following error message:

```console
1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Waiting 1 minute then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
```

> If the drain continues to fail, the MCC logs the a second error message:

Reviewer: ```suggestion
If the drain continues to fail, the MCC logs a second error message:
```

```

After one hour has passed and the drain is still failing, the MCC logs the following error messages and the node is marked degraded.

```console
1 drain_controller.go:352] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: drain exceeded timeout: 1h0m0s. Will continue to retry.
```

```console
1 status.go:126] Degraded Machine: xxxxxxx-xxxxxx-xxxxxx-xxxxx: and Degraded Reason: failed to drain node: xxxxxxx-xxxxxx-xxxxxx-xxxxx after 1 hour. Please see machine-config-controller logs for more information
```

The MCC logs explicitly list which pods are failing to drain. You should start examining the listed pods for the problem that is causing the drain to fail. Common reasons why a node cannot drain a pod include the following conditions:

- The pod has a PodDisruptionBudget (PBD) that prevents the MCO/MCC from deleting the pod.
- The pod has storage attached and the kubelet is unable to unmount the storage.
- The pod has a webhook that is configured to target UPDATE operations or the webhook is not being called by the kube-apiserver.
- The pod has finalizers set in the pod that are preventing it from terminating.

## Mitigation 

> - If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the previous value after the upgrade completes.
> 
> ```console
> oc patch pdb $PDB -n $NS --type=merge -p '{"spec":{"minAvailable":0}}'

Reviewer: ```suggestion
- If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the previous value after the upgrade completes.
+
```console
oc patch pdb $PDB -n $NS --type=merge -p '{"spec":{"minAvailable":0}}'
```

```
- If a webhook is preventing the pod deletion:

Check the `openshift-kube-apiserver` logs to see exactly what webhook is preventing deletion.

```console
Failed calling webhook, failing open vault.hashicorp.com: failed calling webhook "webhook_name": failed to call webhook: Post "https://hashi-vault-agent-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled2022-10-11T19:13:53.345381984Z E1011 19:13:53.345348      16 dispatcher.go:184] failed calling webhook "webhook_name": failed to call webhook: Post "https://name-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled

> ```

Reviewer: Add a `+` here

Check if it's a `mutating` or `validating` webhook.

> ```console

Reviewer: Add a carriage return and  `+` here

$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration

> ```

Reviewer: Add a carriage return and  `+` here

> $ oc get mutatingwebhookconfiguration
> ```
> Backup and delete the webhook 

Reviewer: ```suggestion
$ oc get mutatingwebhookconfiguration
```
+
Backup and delete the webhook 
```

```console
$ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
$ oc delete <webhook_type> <webhook_name>         

> ```

Reviewer: Add a carriage return and  `+` here

This should allow the drain to continue and if the webhook does not come back automatically you can recreate it via the backup.

- If a pod cannot unmount storage, troubleshoot why it's failing. For example, if you are using NFS storage, the problem could be a network issue with the storage server. 

> 

Reviewer: Add a  `+` here

Otherwise, if you are comfortable with possible data loss, you can force delete the pod and immediately remove resources from the API and bypass graceful deletion: 

```console
$ oc delete pod <pod-name> --force=true --grace-period=0
```
- If the finalizers are causing the pod to be stuck in terminating status you can try patching the finalizers to null:

> 

Reviewer: Add a  `+` here

```console
oc patch pod <pod_name> -p '{"metadata":{"finalizers":null}}' -n <namespace_name>

> ```

Reviewer: Add a carriage return and + here

# MCCDrainError

## Meaning 

> Alerts the user to when the Machine Config Operator 

Reviewer: ```suggestion
Alerts the user to when the Machine Config Operator (MCO)
```

Author: Agreed, Applied the change

> failed to drain a node. Preventing the MCO from restarting 

Reviewer: Somehow, putting this in present tense seems better to me. A purely optional suggestion.
```suggestion
fails to drain a node, which prevents the MCO from restarting 
```

Author: Agreed, Applied the change

the node. 

>Alerts the user to a failed node drain. Always triggers when the failure
>happens one or more times.

This alert will fire as a warning when: 

> - If the mco is unable to drain the node for an hour
> - When the failure occurs one or more times in a node 

Reviewer: ```suggestion
- The MCO is unable to drain the node for an hour
- The failure occurs one or more times in a node 
```

Author: Agreed, Applied the change

## Impact 

> If the mco fails to drain a node it will be unable to reboot the node. 

Reviewer: ```suggestion
If the MCO fails to drain a node it will be unable to reboot the node. 
```

> If the mco fails to drain a node it will be unable to reboot the node. 
> Preventing any changes via MachineConfig or a cluster upgrade to be applied. 

Reviewer: ```suggestion
If the MCO fails to drain a node, it will be unable to reboot the node, which prevents any changes to the cluster through an machine config and prevents a cluster upgrade. 
```

Author: Agreed, Applied the change

> This will prevent a cluster upgrade from progressing/completing.

Reviewer: I this ^^ sentence redundant? If not:
```suggestion
If the MCO fails to drain a node during an upgrade, the upgrade will not be able to progress/complete.
```

## Diagnosis 

> If a node is failing to draining the first place the alert itself recommends to check the machine-config-controller pod inside the openshift-machine-config-operator namespace. 

Reviewer: It seems that other runbooks mark-up Kubernetes objects like these.
```suggestion
If a node fails to drain, first check the `machine-config-controller` pod inside the `openshift-machine-config-operator` namespace. 
```

The machine-config-controller pod is the central point of management for incoming updates to 

> machines.

Author: Thanks for letting me know on this one. Applied the change

> For the commands below replace the $CONTROLLERPOD variable with the name of your own machine-config-controller pod name. 

Reviewer: It is preferred not to use directional words like below and above.
```suggestion
For the following commands, replace the $CONTROLLERPOD variable with the name of your own `machine-config-controller` pod name. 
```

Author: Agreed, Applied the change

```console
oc -n openshift-machine-config-operator logs $CONTROLLERPOD -c machine-config-controller
```

> When a drain is initiated the following log will be recorded.

Reviewer: Make this past tense. For a second, I thought the previous command triggered the drain (d'oh).
```suggestion
When the MCO drains a node, the following log entry is recorded:
```

Author: Agreed, Applied the change

```console
1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: initiating drain
```

> The controller will then log each and every pod drained from the node. 

Reviewer: For some reason unknown to me, the end user docs don't like the work _will_.
```suggestion
The MCC then logs the name of each pod that is drained from the node. 
```

Author: Agreed, Applied the change

```console
  1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Evicted pod <namespace>/<pod-name>
```

> If the controller is unable to drain a pod after 1m30s then it will fire an error log.

Reviewer: Does the MCC do the drain?
```suggestion
If the MCO/MCC is unable to drain a pod after 1m30s, it logs the following error message:
```

Author: As far as I know yes, the machine config controller is what initiates and does drains for nodes but we can just specify MCO/MCC as a catch all

```console
1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Waiting 1 minute then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
```

> If the drain continues to fail then another failure log will fire.

Author: Agreed, Applied the change

```console
 1 drain_controller.go:173] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: Drain failed. Drain has been failing for more than 10 minutes. Waiting 5 minutes then retrying. Error message from drain: error when waiting for pod "xxxx-xxxx-xx" in namespace "xxxxxxx" to terminate: global timeout reached: 1m30s
```

> Finally if 1 hour has passed and the drain is still failing the following logs will be fired and the node will be marked degraded.

Reviewer: We spell out numbers lower than 10.
```suggestion
After one hour has passed and the drain is still failing, the MCC logs the following error messages and the node is marked degraded.
```

Author: Agreed, Applied the change

```console
1 drain_controller.go:352] node xxxxxxx-xxxxxx-xxxxxx-xxxxx: drain exceeded timeout: 1h0m0s. Will continue to retry.
```

```console
1 status.go:126] Degraded Machine: xxxxxxx-xxxxxx-xxxxxx-xxxxx: and Degraded Reason: failed to drain node: xxxxxxx-xxxxxx-xxxxxx-xxxxx after 1 hour. Please see machine-config-controller logs for more information
```

> The machine-controller logs will explicitly tell you what pods are failing to drain succesfully. Looking at the pods named in the logs is where you should start. 

Reviewer: ```suggestion
The MCC logs explicitly list which pods are failing to drain. You should start examining the listed pods for the problem that is causing the drain to fail. Common reasons why a node cannot drain a pod include the following conditions:
```

Author: Agreed, Applied the change

> Common reasons a node cannot drain a pod successfully are:

Author: Agreed, Applied the change

> - The pod has a PodDisruptionBudget (PBD) set preventing the controller from deleting the pod.
> - The pod has storage attached and the kubelet is unable to unmount the storage succesfully.
> - The pod deletion may be prevented by a webhook if it's configured to target UPDATE operations or the webhook is failing to be called by the kube-apiserver.
> - The pod deletion could be stuck because the finalizers set in the pod are preventing it from terminating succesfully.

Reviewer: ```suggestion
- The pod has a PodDisruptionBudget (PBD) that prevents the MCO/MCC from deleting the pod.
- The pod has storage attached and the kubelet is unable to unmount the storage.
- The pod has a webhook that is configured to target UPDATE operations or the webhook is not being called by the kube-apiserver.
- The pod has finalizers set in the pod that are preventing it from terminating.
```

Author: Agreed, Applied the change

## Mitigation 

> - If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the old value after the upgrade completes.

Reviewer: ```suggestion
- If a PDB is causing the failure you can temporarily patch it to set the `minAvailable` to 0 so it can scale the pod down successfully. Then patch it to the previous value after the upgrade completes.
```

Author: Agreed, Applied the change

```console
oc patch pdb $PDB -n $NS --type=merge -p '{"spec":{"minAvailable":0}}'

> ```

Reviewer: Should we make the following steps into a procedure?

Author: I'm hesitant on this one. Do you mean a more detailed procedure such as scaling to 0 then back up?

Reviewer: I was thinking something like:

If a webhook is preventing the pod deletion:
1. Check the `openshift-kube-apiserver` logs 
 Failed calling webhook....
 
2. Check if it's a `mutating` or `validating` webhook.
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration

3. Backup and delete the webhook
$ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
$ oc delete <webhook_type> <webhook_name>

Author: Oh you mean for the webhook? The PDB and webhook are separate points. I can add step numbers to the webhook if you think it's needed. But the PDB and webhook are not together that's why I have them marked with a - as they are different points.

Reviewer: Actually, after looking at the [View File preview](https://github.com/openshift/runbooks/blob/7322f7892f7c57c9db08c0df54c0e7d904220107/alerts/machine-config-operator/MachineConfigControllerDrainError.md#mitigation), I think the Mitigation section is perfectly clear and easy to follow as you have it now. 
Just need some `+` characters to tuck the code blocks and text under the bulleted lines.

Reviewer: @CourtCourt521 I see you used spaces to line up the steps and its code block. In docs we use a `+`. Does the spacing render properly?

. Check if it's a `mutating` or `validating` webhook.
+
 ```console
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration
```

Author: Hi,

No when I do it with my code it uses the + as a bullet point. It looks like this. 

- Check if it's a mutating or validating webhook.
+ 
```console
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration
```

Author: @mburke5678 ^

> - If a webhook is preventing the pod deletion you can check the `openshift-kube-apiserver` logs to see exactly what webhook is preventing deletion.

Reviewer: ```suggestion
- If a webhook is preventing the pod deletion you can check the `openshift-kube-apiserver` logs to see which webhook is preventing deletion.
```

> - If a webhook is preventing the pod deletion you can check the `openshift-kube-apiserver` logs to see exactly what webhook is preventing deletion.
> 
> ```console
> Failed calling webhook, failing open vault.hashicorp.com: failed calling webhook "webhook_name": failed to call webhook: Post "https://hashi-vault-agent-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled2022-10-11T19:13:53.345381984Z E1011 19:13:53.345348      16 dispatcher.go:184] failed calling webhook "webhook_name": failed to call webhook: Post "https://name-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled
> ```
> Then check if it's a `mutating` or `validating` webhook.
> ```console
> $ oc get validatingwebhookconfiguration
> $ oc get mutatingwebhookconfiguration
> ```
> Then backup and delete the webhook 
> ```console
> $ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
> $$ oc delete <webhook_type> <webhook_name>         
> ```
> This should allow the drain to continue and if the webhook does not come back automatically you can recreate it via the backup.

Reviewer: I can't get the formatting to look right in the comment preview, but you get the idea:
```suggestion
- If a webhook is preventing the pod deletion:
+
. Check the `openshift-kube-apiserver` logs to see exactly what webhook is preventing deletion.
+
```console
Failed calling webhook, failing open vault.hashicorp.com: failed calling webhook "webhook_name": failed to call webhook: Post "https://hashi-vault-agent-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled2022-10-11T19:13:53.345381984Z E1011 19:13:53.345348      16 dispatcher.go:184] failed calling webhook "webhook_name": failed to call webhook: Post "https://name-injector-svc.vault-injector.svc:443/mutate?timeout=30s": context canceled
```
+
. Check if it's a `mutating` or `validating` webhook.
+
```console
$ oc get validatingwebhookconfiguration
$ oc get mutatingwebhookconfiguration
```
+
. Backup and delete the webhook: 
+
```console
$ oc get validatingwebhookconfiguration/<webhook_name> -o yaml > webhook.yaml
$$ oc delete <webhook_type> <webhook_name>         
```
+
This should allow the drain to continue and if the webhook does not come back automatically you can recreate it via the backup.
```

Author: Agreed, Applied the change

> - If a pod cannot amount storage you either need to troubleshoot why it's failing such as if it's a NFS storage then it could be a network issue with the storage server. If you are okay with possible data loss you can force delete the pod and immediately remove resources from the API and bypass graceful deletion. 

Reviewer: ```suggestion
- If a pod cannot unmount storage, troubleshoot why it's failing. For example, if you are using NFS storage, the problem could be a network issue with the storage server. 

Otherwise, if you are comfortable with possible data loss, you can force delete the pod and immediately remove resources from the API and bypass graceful deletion by using the following command: 
```

Author: Agreed, Applied the change

```console
$ oc delete pod <pod-name> --force=true --grace-period=0
```

> - If the pod is stuck in terminating status you can try patching the finalizers to null.

Author: Agreed, Applied the change

```console
oc patch pod <pod_name> -p '{"metadata":{"finalizers":null}}' -n <namespace_name>
```

> or you can try force deleting the pod as described above. 

Reviewer: ```suggestion
Or, you can force delete the pod using the previous command. 
```

Author: Agreed, Applied the change
